core/src/main/scala/zquery/DataSource.scala

```diff
@@ -0,0 +1,149 @@
+     * that this data source can execute.
+     */
+    final def contramapM[R1 <: R, E1 >: E, B](name: String)(f: B => ZIO[R1, E1, A]): DataSource.Service[R1, E1, B] =
+      new DataSource.Service[R1, E1, B] {
+        val identifier = s"${self.identifier}.contramapM($name)"
+        def run(requests: Iterable[B]): ZIO[R1, E1, CompletedRequestMap] =
+          ZIO.foreach(requests)(f).flatMap(self.run)
+      }
+
+    /**
+     * Returns a new data source that executes requests of type `C` using the
+     * specified function to transform `C` requests into requests that either
+     * this data source or that data source can execute.
+     */
+    final def eitherWith[R1 <: R, E1 >: E, B, C](
+      that: DataSource.Service[R1, E1, B]
+    )(name: String)(f: C => Either[A, B]): DataSource.Service[R1, E1, C] =
+      new DataSource.Service[R1, E1, C] {
+        val identifier = s"${self.identifier}.eitherWith(${that.identifier})($name)"
+        def run(requests: Iterable[C]): ZIO[R1, E1, CompletedRequestMap] = {
+          val (as, bs) = requests.foldLeft((List.empty[A], List.empty[B])) {
+            case ((as, bs), c) =>
+              f(c) match {
+                case Left(a)  => (a :: as, bs)
+                case Right(b) => (as, b :: bs)
+              }
+          }
+          self.run(as).zipWithPar(that.run(bs))(_ ++ _)
+        }
+      }
+
+    override final def equals(that: Any): Boolean = that match {
+      case that: DataSource.Service[_, _, _] => this.identifier == that.identifier
+    }
+
+    override final def hashCode: Int =
+      identifier.hashCode
+
+    /**
+     * Returns a new data source with failures mapped using the specified
+     * function.
+     */
+    final def mapError[E1](name: String)(f: E => E1): DataSource.Service[R, E1, A] =
+      new DataSource.Service[R, E1, A] {
+        val identifier = s"${self.identifier}.mapError($name)"
+        def run(requests: Iterable[A]): ZIO[R, E1, CompletedRequestMap] =
+          self.run(requests).mapError(f)
+      }
+
+    /**
+     * Provides this data source with its required environment.
+     */
+    final def provide(name: String)(r: R): DataSource.Service[Any, E, A] =
+      provideSome(s"_ => $name")(_ => r)
+
+    /**
+     * Provides this data source with part of its required environment.
+     */
+    final def provideSome[R0](name: String)(f: R0 => R): DataSource.Service[R0, E, A] =
+      new DataSource.Service[R0, E, A] {
+        val identifier = s"${self.identifier}.provideSome($name)"
+        def run(requests: Iterable[A]): ZIO[R0, E, CompletedRequestMap] =
+          self.run(requests).provideSome(f)
+      }
+
+    override final def toString: String =
+      identifier
+  }
+
+  /**
+   * Constructs a data source from a pure function.
+   */
+  final def fromFunction[A, B](
+    name: String
+  )(f: A => B)(implicit ev: A <:< Request[B]): DataSource.Service[Any, Nothing, A] =
+    new DataSource.Service[Any, Nothing, A] {
```

Comment:
How about returning `DataSource` directly?

Reply:
I changed `fromRequestWith` to accept a `DataSource.Service` instead of a `DataSource` that way all the methods to create data sources and combinators for them return the service and you only have to deal with the module if you want to use a data source as part of the environment.